multi-node DDP support

sqchai opened the pull request:
1. Fixed 'store_snapshot'. Previously each node will try to write a snapshot to the same root directory in multi-node multi-gpu mode. Restored 'store_snapshot' default value to True. Modified such that only node rank = 0 process will write a snapshot.
2. Previous change to train.py on line 255 to line 311 (multi-gpu part) is not necessary. Restored to original code.
3. For the 'disallow_scheduler' part in 'process_environment.py'. For some reasons my previous local code base do not have this 'disallow_scheduler' function, and does not import this function in the script. Therefore, this part of code always fails. I fixed those things according to the latest alf implementation.
4. Modified README.md to include instructions to launch multi-node multi-gpu training.
5. Fixed code style problem.

Commit:
multi-nodes multi-gpus training

Commit:
"Addressed review comments"

JingyuQian commented on alf/bin/train.py:
```diff
@@ -176,13 +179,12 @@ def training_worker(rank: int,
             # Specialization for distributed mode
             dist.init_process_group('nccl', rank=rank, world_size=world_size)
             # Recover the flags when spawned as a sub process
-            if rank > 0:
```
this `if` condition, if present during "multi-gpu" mode, will raise a `"root_dir" has been defined twice` error. Removing it seems to solve the problem. Need further confirmation?

emailweixu replied:
On the contrary, `if` condition is needed so that flags will not be redefined for rank0 because rank0 process is the main process, which already defined the flags.

Commit:
add new line at end of file

Commit:
change new line

Commit:
fixed doc str

emailweixu commented on alf/bin/train.py:
```diff
@@ -215,7 +279,7 @@ def main(_):
     conf_file = common.get_conf_file()
 
-    if FLAGS.store_snapshot:
+    if FLAGS.store_snapshot and int(os.environ['RANK']) == 0:
```
There may not be RANK environment variable for single process / single node job.

emailweixu commented on alf/bin/train.py:
```diff
@@ -264,6 +328,40 @@ def main(_):
             raise ChildProcessError(f'Training failed on subprocess exception')
 
+    elif FLAGS.distributed == 'multi-node-multi-gpu':
+        local_rank = int(os.environ['LOCAL_RANK'])
```
why not use FLAGS.local_rank?

sqchai replied:
The 'LOCAL_RANK' is automatically assigned by PyTorch launcher. I'm not sure how to pre-assign this from a shell script

emailweixu commented on alf/bin/train.py:
```diff
@@ -264,6 +328,40 @@ def main(_):
 
+    elif FLAGS.distributed == 'multi-node-multi-gpu':
+        local_rank = int(os.environ['LOCAL_RANK'])
+        rank = int(os.environ['RANK'])
+        world_size = int(os.environ['WORLD_SIZE'])
```
And I couldn't find documentation about RANK or WORLD_SIZE environment variable for torch.distributed.launch. Can you give the link for the documentation?

sqchai replied:
This is updated in the new README

emailweixu commented on alf/utils/per_process_context.py:
```diff
@@ -78,6 +81,10 @@ def is_distributed(self):
         return self._ddp_rank
 
+    @property
+    def local_rank(self):
```
since now there are two rank, ddp_rank and local_rank, it is better to clearly document what their meanings are.

emailweixu commented on alf/utils/spawned_process_utils.py:
```diff
@@ -31,6 +31,7 @@ class SpawnedProcessContext(NamedTuple):
     """
     ddp_num_procs: int
     ddp_rank: int
+    local_rank: int
```
comment ddp_rank and local_rank

emailweixu commented on alf/bin/train.py:
```diff
@@ -264,6 +328,40 @@ def main(_):
+            # in different work processes.
+            manager = mp.Manager()
+            paras_queue = manager.Queue()
+            training_worker_multi_node(
```
I guess Queue does not work across multiple machines. If so,  need to update Trainer._check_dpp_paras_consistency for multi-node mode. Perhaps change `if self._rank > 0:` to `if local_rank > 0:`. If so, need to update the doc-string for TrainerConfig.ddp_paras_check_interval to say that it only check the consistency within one node. And please test if TraomerConfig.check_dpp_paras_consistency=True works properly.

emailweixu replied:
Actually, I think `paras_queue` doesn't work at all. In multi-gpu mode, `paras_queue` can work because it is created by the main process. All the worker will use the same paras_queue. Here different worker will use the different queue, which makes no sense.
So it's better just to report unsupported error for TraomerConfig.check_dpp_paras_consistency=True in multi-node-multi-gpu mode.

emailweixu replied:
Can you address this?

sqchai replied:
I'm currently addressing this in policy_trainer.py. I'm not sure how to get this flag value in train.py.  It seems like the config is not yet parsed from file in train.py?

emailweixu commented on alf/bin/train.py:
```diff
@@ -191,7 +194,68 @@ def training_worker(rank: int,
+    try:
+        _setup_logging(log_dir=root_dir, rank=rank)
+        _setup_device(local_rank)
+        if world_size > 1:
```
it's guaranteed that `world_size > 1`, right? So there is no need for the `if`

emailweixu commented on alf/bin/train.py:
```diff
@@ -133,6 +135,8 @@ def _train(root_dir, rank=0, world_size=1):
 
     if trainer_conf.ml_type == 'rl':
         ddp_rank = rank if world_size > 1 else -1
+        if ddp_rank > -1 and local_rank > -1:
+            ddp_rank = local_rank
```
what is the reason for this?

Commit:
Addressed review comments

emailweixu commented:
Need resolve conflict first.

sqchai committed:
resolve conflict

emailweixu commented:
Still need to resolve conflict

sqchai committed:
doc string

sqchai committed:
Update train.py

sqchai committed:
spelling

emailweixu commented on alf/bin/train.py:
```diff
@@ -190,6 +203,13 @@ def _train(root_dir, rank=0, world_size=1):
     trainer.train()
 
 
+def _training_worker_helper(rank: int, *args, **kwargs):
+    # Helper to start the training worker with the correct rank
+    # so that rank 0 is from the main process and the rest are
+    # from the spawned processes.
+    training_worker(rank + 1, *args, **kwargs)
```
no longer used.

emailweixu commented on alf/bin/train.py:
```diff
@@ -147,12 +159,13 @@ def _setup_remote_configs_if_needed():
 
 
-def _train(root_dir, rank=0, world_size=1):
+def _train(root_dir, local_rank=-1, rank=0, world_size=1):
```
local_rank is not used.

emailweixu commented on alf/trainers/policy_trainer.py:
```diff
@@ -808,6 +808,12 @@ def _check_dpp_paras_consistency(self, iter_num: int,
             return
 
         proc_cxt = PerProcessContext()
+        if (proc_cxt.is_distributed and proc_cxt.local_rank >= 0
```
This error can be reported earlier in train.py

sqchai committed:
changed ddp consistency flag check location

Commit:
fix

Commit:
yapf

sqchai committed:
switch to torchrun

Commit:
yapf

sqchai commented:
1. Switched the launching method to torchrun
2. The setup_device() function was modified in a previous commit to support Taichi, which caused problems to multi-node DDP launching. I updated the multi-node training worker so they're compatible now.

emailweixu commented on alf/bin/train.py:
```diff
@@ -428,13 +428,20 @@ def main(_):
             manager = mp.Manager()
             paras_queue = manager.Queue()
+            CUDA_VISIBLE_DEVICES = os.environ.get('CUDA_VISIBLE_DEVICES')
+            os.environ['CUDA_VISIBLE_DEVICES'] = str(local_rank)
```
If the original CUDA_VISIBLE_DEVICES is set, it needs to be set based on the original CUDA_VISIBLE_DEVICES here, similar to what single node DDP does.

sqchai replied:
Yes I tried this, but If the CUDA_VISIBLE_DEVICES is set to 2,3, and the local_rank will be assigned as logical rank 0,1. In this case, using the physical device id can cause problems.

emailweixu replied:
> Yes I tried this, but If the CUDA_VISIBLE_DEVICES is set to 2,3, and the local_rank will be assigned as logical rank 0,1. In this case, using the physical device id can cause problems.

What is the problem? If CUDA_VISIBLE_DEVICES is not supported, it should report error.

Commit:
visible devices
